### DISCUSSION TITLE
Towards a 'language of (software) design'

### DISCUSSION THREAD

**Discussion by JacquesCarette** [JacquesCarette/Drasil#3313] [Ideas]
(This is in Discussions for now, as it is really a pre-project, i.e. it is much too early to implement anything. It's also too early to really start a wiki page, as the content is a starting point for an investigation project, which would itself be preparation for a potential implementation project.)

One of the benefits of Drasil is that it gives us a clear place to create a "language of (software) design". Of course, we're not the first to think about such a thing! However previous work doesn't seem to have quite had a sufficient impact that any of it is well-known. Contrast this to, for example, software patterns, which are.

So we need to explore the previous work, as well as sources of inspiration. Here, I'm going to put various 'leads'. 

Wikipedia:
- [Software Design](https://en.wikipedia.org/wiki/Software_design) . If this were all a 'science', our answers would be here already. Note well the ToC sidebar, which leads to lots of related material.
- [Software Construction](https://en.wikipedia.org/wiki/Software_construction) is also related since the construction is the action of taking the design (decisions) and making them into an actual thing. Just having a language of design is not enough, it needs to be **interpretable** as a method of construction.

Places we could learn from:
- modeling languages
- design patterns
- "systems design" (i.e. outside software)
- refactoring (the paper [The untold story of code refactoring customizations in practice](https://customrefactoring.github.io/download/preprint.pdf) is what prompted me to start writing all of this down). I think we can reverse-engineer ideas of design from the refactorings that people do.
- The essay [Software Metaphors](https://www.researchgate.net/publication/265757406_Software_Metaphors) (which seems to not be published)
- [FOAM3](https://github.com/kgrgreer/foam3)

Outside of software, design also exists (of course). The useful part: they have languages!
- on wikipedia [Design Language](https://en.wikipedia.org/wiki/Design_language) mostly talks about 'industrial design'.

There's surely a lot more, I'll update the discussion as I find more.

**Comment by smiths** [comment]
Great discussion topic @JacquesCarette.  I clicked through your links and I believe I understand the presence of all of them.  However, the last one ([Design Language](https://en.wikipedia.org/wiki/Design_language)) feels like an outlier.  It seems like it is just about the aesthetics of a product, not about its functionality.  Maybe if I tracked downs some of the examples design languages, at least the more rigorous ones, I might get a better feel for the relevance?

The Design Language page does have links to [Architecture Description Languages](https://en.wikipedia.org/wiki/Architecture_description_language) and [Unified Modeling Language](https://en.wikipedia.org/wiki/Unified_Modeling_Language).  Those seem like relevant links for a language of software design.

**Reply by JacquesCarette** [reply]
Well spotted, and that gives me an opportunity to explain something that was in my mind, but didn't explain above.

What if you've implemented all of the functional requirements of a piece of software, but none of its non-functional ones? Are you now stuck? In particular, have you run out of 'language of design'? Are the actions needed to also deal with all the non-functional requirements not 'design'?

Of course they are. They also involve code, making non-trivial decisions, satisfying constraints, etc. It's mostly that testing for NFRs is relational rather than functional.

The same is true for various kinds of aesthetic design of products: it's about fulfilling a different set of requirements, which are indeed non-functional ones. But, as many have shown, and Apple quite spectacularly so, that kind of 'design' matters a lot (in some circumstances). And, lo and behold, that kind of design, as well as graphic design, information display design, and so on, all have languages too.

So rather than deleting that link, I should greatly expand on it.
